ethereum/core/src/main/java/org/hyperledger/besu/ethereum/chain/DefaultBlockchain.java

```diff
@@ -386,16 +416,19 @@ private BlockAddedEvent handleChainReorg(
     // Update tracked forks
     final Collection<Hash> forks = blockchainStorage.getForkHeads();
     // Old head is now a fork
-    forks.add(oldChainHead);
+    forks.add(oldChainHeadBlock.getHash());
     // Remove new chain head's parent if it was tracked as a fork
     final Optional<Hash> parentFork =
-        forks.stream().filter(f -> f.equals(newChainHead.getHeader().getParentHash())).findAny();
+        forks.stream()
+            .filter(f -> f.equals(newChainHeadWithReceipts.getHeader().getParentHash()))
+            .findAny();
     parentFork.ifPresent(forks::remove);
     updater.setForkHeads(forks);
     return BlockAddedEvent.createForChainReorg(
```

Comment:
Can we add tests that check that the logs we're emitting here are what we expect and are ordered correctly?